salvo/src/lib/benchmark/fully_dockerized_benchmark.py

```diff
@@ -24,25 +24,132 @@ def __init__(self, job_control: JobControl, benchmark_name: str) -> None:
     """Validate that sources are available to build a missing image.
 
+    Verify that a source definition exists tht can build a missing
+    image needed for the benchmark.
+
     Returns:
         None
+
+    Raises:
+        an Exception if no source definitions allow us to build
+          missing docker images.
     """
-    pass
+    source = self.get_source()
+    if not source:
+      raise Exception("No source configuration specified")
+
+    can_build_envoy = False
+    can_build_nighthawk = False
+
+    for source_def in source:
+      # Cases:
+      # Missing envoy image -> Need to see an envoy source definition
+      # Missing at least one nighthawk image -> Need to see a nighthawk source
+
+      if source_def.identity == source_def.SRCID_UNSPECIFIED:
+        raise Exception("No source identity specified")
+
+      if not images.envoy_image \
+          and source_def.identity == source_def.SRCID_ENVOY and \
+          (source_def.source_path or source_def.source_url):
+        can_build_envoy = True
+
+      if (not images.nighthawk_benchmark_image or not images.nighthawk_binary_image) \
+          and source_def.identity == source_def.SRCID_NIGHTHAWK and \
+          (source_def.source_path or source_def.source_url):
+        can_build_nighthawk = True
+
+      if (not images.envoy_image and not can_build_envoy) or \
+          (not images.nighthawk_benchmark_image or not images.nighthawk_binary_image) \
+              and not can_build_nighthawk:
+        # If the Envoy image is specified, then the validation failed for
+        # NightHawk and vice versa
+        msg = "No source specified to build unspecified {image} image".format(
+            image="NightHawk" if images.envoy_image else "Envoy")
+        raise Exception(msg)
+
 
   def execute_benchmark(self) -> None:
     """Prepare input artifacts and run the benchmark.
 
+    Construct the volume, environment variables, and command line
+    arguments needed to execute the benchmark image.
+
     Returns:
         None
+
+    Raises:
+      NotImplementedError if the benchmark is configured to execute
+        remotely.
     """
     self._validate()
+
+    if self.is_remote():
+      raise NotImplementedError("Local benchmarks only for the moment")
+
+    # pull in environment and set values
+    output_dir = self._control.environment.output_dir
+    test_dir = self._control.environment.test_dir
+    images = self.get_images()
+    log.debug(f"Images: {images.nighthawk_benchmark_image}")
+
+    # 'TMPDIR' is required for successful operation.
+    image_vars = {
+        'NH_DOCKER_IMAGE': images.nighthawk_binary_image,
+        'ENVOY_DOCKER_IMAGE_TO_TEST': images.envoy_image,
+        'TMPDIR': output_dir
+    }
+    log.debug(f"Using environment: {image_vars}")
+
+    volumes = get_docker_volumes(output_dir, test_dir)
+    log.debug(f"Using Volumes: {volumes}")
+    self.set_environment_vars()
+
+    kwargs = {}
+    kwargs['environment'] = image_vars
```

Comment:
This looks like a good time to convert the `kwargs` to a more structured and better documented parameter to `run_image()`.

We can use something very similar to `kwargs`, but with a fixed and documented structure a `namedtuple`. The advantage is that the structure becomes documented and is easier to understand. There are two ways to define and document a namedtuple, feel free to choose based on your preference.

```python
import collections

class LocalRunParameters(collections.namedtuple("LocalRunParameters", "command volumes network_mode tty")):
  """Describes parameters of a local execution.

  Attributes:
    command: A string, ...
    volumes: ...
    network_mode: ...
    tty: ...
  """

or this:

# Describes parameters of an animal.
LocalRunParameters = collections.namedtuple('LocalRunParameters', [
    'command',  # A string, ...
    'volumes',    # ...
    'network_mode',      # ...
    'tty', # ...
])
```

The problem with plain `kwargs` is that the reader of the code has absolutely no idea what parameters are supported or required, nor what is their meaning.

Reply:
I agree about `kwargs`. Let me look into using this method

Reply:
That looks great, can we also add a top level comment above all the locations where we defined a `namedtuple`, since these are public classes in essence?

Example for `DockerRunParameters`
```python
# Provides parameters used when running benchmark in docker.
DockerRunParameters = collections.namedtuple("DockerRunParameters", [
    'environment',  # a dict with environment variable to set in the container
    'command',      # a string containing the command to execute
    'volumes',      # a dict with the volumes mounted in the container
    'network_mode', # a string that specifies the network stack used
    'tty'           # a boolean indicating if a pseudo-tty is allocated
])
```

(optional) Consider adding `pytype` annotations to the namedtuple, see `typing.NamedTuple` which should work just like collections.namedtuple:

```python
Point = typing.NamedTuple("Point", [('x', int), ('y', int)])
```

Reply:
I was able to get this going for one of the `NamedTuple` instances.  I'll add a `#TODO` and work on the typing later.  `Dicts` seem to be troublesome